### DISCUSSION TITLE
How to change default behavior for Node on Enter key?

### DISCUSSION THREAD

**Discussion by emma-campbell** [ueberdosis/tiptap#2176] [General Questions]
Hello 👋🏼 I've been trying to implement an editor for an internal tool for my company, and need to force a placeholder for a title and subtitle. Both of these may be left blank. After the subtitle, blocks of content. This is the structure that I've come up with.

```typescript
import { Editor, EditorContent } from '@tiptap/vue-2';
import { Node } from '@tiptap/core';
import Document from '@tiptap/extension-document';
import Placeholder from '@tiptap/extension-placeholder';

const Title = Node.create({
  name: 'title',
  content: 'inline*',
  parseHTML() {
    return [{ tag: 'h1' }];
  },
  renderHTML({ HTMLAttributes }) {
    return ['h1', HTMLAttributes, 0];
  }
});

const Subtitle = Node.create({
  name: 'subtitle',
  content: 'inline*',
  parseHTML() {
    return [{ tag: 'h2' }];
  },
  renderHTML({ HTMLAttributes }) {
    return ['h2', HTMLAttributes, 0];
  }
});

const Content = Node.create({
  name: 'content',
  content: 'block+',
  parseHTML() {
    return [{ tag: 'div' }];
  },
  renderHTML({ HTMLAttributes }) {
    return ['div', HTMLAttributes, 0];
  }
});

const CustomDocument = Document.extend({
  content: 'title subtitle content'
});

export default {
  data() {
      return {
          editor: null,
      }
  },
  mounted() {
    this.editor = new Editor({
      content: this.text,   // computed field that wraps text fields in html
      extensions: [
        CustomDocument,
        Title,
        Subtitle,
        Placeholder.configure({
          placeholder: ({node}) => {
            if (node.type.name === 'title') {
              return 'Title...';
            } else if (node.type.name === 'subtitle') {
              return 'Subtitle...';
            }
            return 'Write your content here...';
          }
       })
    ]});
  }
}
```
However, I can't seem to get around some unwanted behavior. When I press enter while the cursor is on the title or subtitle, the cursor will not advance to the next line (in this case, node). I am trying to come up with a work around, and thought something similar to #426 would work. Maybe something along the lines of...

```typescript
const GoToNextNode = Extension.create({
  keys() {
    return {
      Enter(state, dispatch, view) {
        // if node.type.name == 'title' or node.type.name == 'subtitle'
        // go to the next node
        // else 
        // create a new block node below
      }
    }
  }
});
```

I've read through the [list of commands](https://tiptap.dev/api/commands) and it doesn't look like anything really matches what I am trying to accomplish (maybe `selectNodeForward` or `selectNodeBackward`). I've also been trying to wrap my head around the [editor state](https://prosemirror.net/docs/guide/#state) and transactions, as it seems they will be relevant. However, I am stuck on how I would access the next node and change the state to reflect editing the next node.

Any ideas?

**Comment by codemzy** [comment]
Adding the defining and group schema to the heading nodes made this work for me, by default hitting enter goes to the next line/node, without using any work around. I also got rid of the content Node since everything after the titles is the normal editor. (I'm not using Vue in this example, but it should work the same).

```js
const Title = Node.create({
  name: 'title',
  content: 'inline*',
  group: 'block',
  defining: true,
  parseHTML() {
    return [{ tag: 'h1' }];
  },
  renderHTML({ HTMLAttributes }) {
    return ['h1', HTMLAttributes, 0];
  }
});

const Subtitle = Node.create({
  name: 'subtitle',
  content: 'inline*',
  group: 'block',
  defining: true,
  parseHTML() {
    return [{ tag: 'h2' }];
  },
  renderHTML({ HTMLAttributes }) {
    return ['h2', HTMLAttributes, 0];
  }
});

const CustomDocument = Document.extend({
  content: 'title subtitle block*',
});
```

Don't forget to add the StarterKit (or bring in the Text, Paragraph extensions for your editor).

```javascript
const editor = new Editor({
  element: document.querySelector('.editable'),
  extensions: [
    StarterKit.configure({
      document: false,
    }),
    CustomDocument,
    Title,
    Subtitle,
    Placeholder.configure({
      placeholder: ({node}) => {
        if (node.type.name === 'title') {
          return 'Title...';
        } else if (node.type.name === 'subtitle') {
          return 'Subtitle...';
        }
        return 'Write your content here...';
      }
   })
  ],
 //...
```

Hope that helps! Theres an [example here in the docs](https://tiptap.dev/examples/custom-document) which is useful.

**Reply by emma-campbell** [reply]
Hi! Thanks for your reply. Unfortunately, this doesn't _quite_ fit what I am looking accomplish... When I implement what you have I get the behavior demonstrated in this [gif](https://ibb.co/w0rRVzT). If I continue to press enter, I get something like this

![Weird Document Behavior](https://i.ibb.co/5TKqDqz/Screen-Shot-2021-11-19-at-10-57-47-AM.png)

Also, I tried to purposely wrap that content in a `div` to I could match it using regex, and then just strip the tags to get the HTML "content". Backend structure is a plain text heading, plain text subheading, and then "content" which can be marked up using html. None of these are stored with outer tags. However, I could probably come up with some other way of parsing the data if this prevents me from being able to attain the expected behavior.

EDIT: typo 🥴

**Reply by codemzy** [reply]
Hmmm, my bad... I didn't have the placeholder CSS marked up so I didn't spot that!!

There are a couple of things that get me closer to what I think you want to achieve.. removing the headers from the block group seems to stop that behaviour (since they are not allowed in the content), but you can't press enter to get you from the title to the subtitle (you have to press the down arrow).

```javascript
const Title = Node.create({
  name: 'title',
  content: 'inline*',
  defining: true,
  parseHTML() {
    return [{ tag: 'h1' }];
  },
  renderHTML({ HTMLAttributes }) {
    return ['h1', HTMLAttributes, 0];
  }
});

const Subtitle = Node.create({
  name: 'subtitle',
  content: 'inline*',
  defining: true,
  parseHTML() {
    return [{ tag: 'h2' }];
  },
  renderHTML({ HTMLAttributes }) {
    return ['h2', HTMLAttributes, 0];
  }
});

const CustomDocument = Document.extend({
  content: 'title subtitle block*',
});
```

What I did find is this works to force a Title and (optional) Subtitle and the enter works nicely... 

```javascript
const CustomDocument = Document.extend({
  content: 'title subtitle?',
});
```

And that got me thinking about [this page in the docs and maybe multiple editors](https://tiptap.dev/experiments/multiple-editors) might be your best bet? One for the title(s) and one for the content? And putting your toolbar on the content editor since that's the only editor users will be able to change the formatting etc (since the titles are forced)? 

Just an idea.. although there may be a way to achieve what you are after in a single editor, I've only recently started building with TipTap myself!

**Reply by emma-campbell** [reply]
Yeah, I originally had a form set up, but thought it may be more intuitive to represent in one editor since to the user it is essentially one document. Why not define it all in one place? Off the top of my head I think [Notion](https://www.notion.so) accomplishes something similar with its document structure.

I may revert back if I can't get it working the way that I would like. Back to the drawing board 🤔
